### PR TITLE
Add support for using the hcode repo

### PR DESCRIPTION
**Pull request by rjknight**
-Add hcode package to enable compiling hardware
     reference image
    -Modify pnor and sbe makefiles to include a dependency on
     the hcode repository
    -Modify config files so that p9 platfoms use the compiled
     version of the hardware reference image by default
    -Update hostboot-binaries.mk to install the ring images
     to the staging dir

### CODE REVIEW

**Comment by stewart-ibm** [comment]
Can one of the admins verify this patch for testing?

**Comment by stewart-ibm** [comment]
```
+ make --directory=/tmp/hostboot-jenkins.swg-devops/jenkins-slave/workspace/Hostboot/OpenPower/GHPR/build-multi/CONFIG/openpower_p9_mambo/OS/rhel6.4/build/op-build/buildroot O=/tmp/hostboot-jenkins.swg-devops/jenkins-slave/workspace/Hostboot/OpenPower/GHPR/build-multi/CONFIG/openpower_p9_mambo/OS/rhel6.4/build/op-build/output
make: Entering directory `/tmp/hostboot-jenkins.swg-devops/jenkins-slave/workspace/Hostboot/OpenPower/GHPR/build-multi/CONFIG/openpower_p9_mambo/OS/rhel6.4/build/op-build/buildroot'
Makefile:542: *** hostboot-binaries is in the dependency chain of hcode that has added it to its _DEPENDENCIES variable without selecting it or depending on it from Config.in.  Stop.
make: *** [_all] Error 2
make: Leaving directory `/tmp/hostboot-jenkins.swg-devops/jenkins-slave/workspace/Hostboot/OpenPower/GHPR/build-multi/CONFIG/openpower_p9_mambo/OS/rhel6.4/build/op-build/buildroot'
```

**Comment by stewart-ibm** [comment]
ok to test

**Comment by rjknight** [comment]
Pushed the wrong branch.. update on the way

**Comment by wghoffa** [comment]
+ ls '/tmp/hostboot-jenkins.swg-devops/jenkins-slave/workspace/Hostboot/OpenPower/GHPR/build-multi-op920/CONFIG/witherspoon/OS/rhel6.4/openpower-ci/patches/op-build/*.patch'
+ echo 'no op-build patches'
no op-build patches
+ '[' 0 -ne 0 ']'
+ op-build_get_pull_request
+ cd /tmp/hostboot-jenkins.swg-devops/jenkins-slave/workspace/Hostboot/OpenPower/GHPR/build-multi-op920/CONFIG/witherspoon/OS/rhel6.4/build/op-build
+ git fetch origin pull/2003/head
fatal: Couldn't find remote ref pull/2003/head
+ exit -1

**Comment by wghoffa** [comment]
98.22346|================================================
 98.22672|Error reported by prdf (0xE500) PLID 0x9000004E
 98.22673|  PRD Signature            : 0x280000 0x42490002
 98.23336|  Signature Description    : pu.obus:k0:n0:s0:p00:c0 (OBPPEFIR[0:3]) PPE general error.
 98.23674|  UserData1   : 0x0028000000000303
 98.23675|  UserData2   : 0x4249000200000000
 98.23675|------------------------------------------------
 98.23677|  Callout type             : Hardware Callout
 98.23843|  CPU id                   : 2061
 98.23844|  Target                   : Physical:/Sys0/Node0/Proc0/OBUS0
 98.23844|  Deconfig State           : DELAYED_DECONFIG
 98.23845|  GARD Error Type          : GARD_Predictive
 98.23845|  Priority                 : SRCI_PRIORITY_MED
 98.23846|------------------------------------------------
 98.23846|  
 98.23847|------------------------------------------------
 98.23848|  Hostboot Build ID: hostboot-b512980/hbicore.bin
 98.23848|================================================
 98.31746|================================================
 98.31747|Error reported by prdf (0xE500) PLID 0x9000004F
 98.31748|  PRD Signature            : 0x280003 0x42490002
 98.31751|  Signature Description    : pu.obus:k0:n0:s0:p00:c3 (OBPPEFIR[0:3]) PPE general error.
 98.31752|  UserData1   : 0x0028000300000303
 98.31752|  UserData2   : 0x4249000200000000
 98.31753|------------------------------------------------
 98.31754|  Callout type             : Hardware Callout
 98.31754|  CPU id                   : 2061
 98.31755|  Target                   : Physical:/Sys0/Node0/Proc0/OBUS3
 98.31756|  Deconfig State           : DELAYED_DECONFIG
 98.31756|  GARD Error Type          : GARD_Predictive
 98.31757|  Priority                 : SRCI_PRIORITY_MED
 98.31758|------------------------------------------------
 98.31758|  
 98.31758|------------------------------------------------
 98.31759|  Hostboot Build ID: hostboot-b512980/hbicore.bin
 98.31760|================================================

**Comment by wghoffa** [comment]
@stewart-ibm - Can you comment on what is failing here? We don't have permissions. This is our last hurdle for content we wanted to get in to the master branch prior to creating our op920 branch.

File: openpower/configs/zz_defconfig [hunk 1]
```diff
@@ -29,6 +29,7 @@ BR2_PACKAGE_IPMITOOL_USB=y
 BR2_PACKAGE_MDADM=y
 BR2_PACKAGE_NCURSES_WCHAR=y
 BR2_PACKAGE_DROPBEAR=y
+BR2_PACKAGE_HCODE=n
```
Comment (wghoffa) [inline openpower/configs/zz_defconfig]: Same comment as openpower/configs/openpower_p9_mambo_defconfig

File: openpower/configs/openpower_p9_mambo_defconfig [hunk 2]
```diff
@@ -26,6 +26,7 @@ BR2_PACKAGE_IPMITOOL_USB=y
 BR2_PACKAGE_MDADM=y
 BR2_PACKAGE_NCURSES_WCHAR=y
 BR2_PACKAGE_DROPBEAR=y
+BR2_PACKAGE_HCODE=n
```
Comment (wghoffa) [inline openpower/configs/openpower_p9_mambo_defconfig]: So even though we don't want to build the HCODE we do want to set BR2_HCODE_INCLUDE_IONV ?

File: openpower/package/hcode/hcode.mk [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+HCODE_DEPENDENCIES = host-binutils host-ppe42-gcc hostboot-binaries
+
+HW_IMAGE_BIN_PATH=output/images/hw_image
+HW_IMAGE_BIN=p9n.hw_image.bin
+HCODE_IMAGE_BIN = p9n.ref_image.bin
+
+CROSS_COMPILER_PATH=$(PPE42_GCC_BIN)
+PPE_TOOL_PATH ?= $(CROSS_COMPILER_PATH)
+PPE_PREFIX    ?= $(PPE_TOOL_PATH)/bin/powerpc-eabi-
```
Comment (wghoffa) [inline openpower/package/hcode/hcode.mk]: powerpc-eabi- with no other chars is intended? I assume so, just want to make sure.

File: openpower/package/hcode/hcode.mk [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+ifeq ($(wildcard /opt/rh/devtoolset-*),)
+CXXPATH= $(dir $(HOSTCXX))
```
Comment (wghoffa) [inline openpower/package/hcode/hcode.mk]: We do we have to explicitly set the CXXPATH in all cases? Isn't it set by default?

File: openpower/package/hcode/hcode.mk (hunk 3 above)
Comment (rjknight) [reply openpower/package/hcode/hcode.mk]: yes, this is just the prefix for finding the tool, it gets pasted on the front of the vars $(CC) $(LD) etc,   which are called powerpc-eabi-cc and so on.

File: openpower/package/hcode/hcode.mk (hunk 4 above)
Comment (rjknight) [reply openpower/package/hcode/hcode.mk]: could be a bad choice of a name I don't see it being used for anything else though, the intent was to grab the path for the local compiler and use it for __EKB_PATH, I think HOSTCXX gets set for all cases by buildroot - but I needed to be able to point to the compiler in the deftoolset dir on pool systems

File: openpower/configs/zz_defconfig (hunk 1 above)
Comment (rjknight) [reply openpower/configs/zz_defconfig]: true, no need for BR2_HCODE_INCLUDE_IONV to be on,  I can fix that up

File: openpower/package/hcode/hcode.mk (hunk 4 above)
Comment (stewart-ibm) [reply openpower/package/hcode/hcode.mk]: This won't go upstream, IBM should fix their systems to not be ancient and broken.

i.e. with what I just merged upstream, you're going to have to manually set PATH on those ancient pool machines. I'll ping folk on internal Slack too.

File: openpower/package/hcode/hcode.mk (hunk 3 above)
Comment (stewart-ibm) [reply openpower/package/hcode/hcode.mk]: I think this is somewhat consistent in how we do other ppe based packages.

File: openpower/configs/openpower_p9_mambo_defconfig (hunk 2 above)
Comment (stewart-ibm) [reply openpower/configs/openpower_p9_mambo_defconfig]: I've regenerated the defconfig files in what I'm merging, we can fix up the polarity of including IONV and have it only for witherspoon in a subsequent patch.

File: openpower/configs/zz_defconfig (hunk 1 above)
Comment (stewart-ibm) [reply openpower/configs/zz_defconfig]: I haven't changed that in what I merged, but we can fix it up in subsequent patch.

**Comment by stewart-ibm** [comment]
Merged with a few fixes as of 3f0e29c5e93cb94a3b19250b3ebca63ffbeb23e7

**Pull request closed without merging**